transaction/std/src/input_materials.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+
+//! Input Materials is a helper struct for the transaction builder.
+//! The transaction builder can get inputs either from input credentials,
+//! or signed contingent inputs. In one case the input is being signed by us,
+//! and in the other it is already signed.
+//!
+//! The transaction builder is required to sort these in order of tx public key
+//! of the first element of the ring being signed, then hand them off to build
+//! the actual signatures. This enum makes it convenient to do this.
+
+use crate::InputCredentials;
+use mc_crypto_keys::CompressedRistrettoPublic;
+use mc_transaction_core::{ring_signature::InputRing, tx::TxIn, Amount, SignedContingentInput};
+
+/// Material that can be used by the transaction builder to create an input to
+/// a transaction.
+#[derive(Debug, Clone)]
+pub enum InputMaterials {
+    /// Signable input materials
+    Signable(InputCredentials),
+    /// Presigned input materials
+    Presigned(SignedContingentInput),
+}
+
+impl From<InputCredentials> for InputMaterials {
+    fn from(src: InputCredentials) -> Self {
+        Self::Signable(src)
+    }
+}
+
+impl From<SignedContingentInput> for InputMaterials {
+    fn from(src: SignedContingentInput) -> Self {
+        Self::Presigned(src)
+    }
+}
+
+impl InputMaterials {
+    /// Get the sort key for whichever type of input this is
+    pub fn sort_key(&self) -> &CompressedRistrettoPublic {
+        match self {
+            InputMaterials::Signable(cred) => &cred.ring[0].public_key,
+            InputMaterials::Presigned(input) => &input.tx_in.ring[0].public_key,
+        }
+    }
+
+    /// Get the amount for whichever type of input this is
+    pub fn amount(&self) -> Amount {
+        match self {
+            InputMaterials::Signable(cred) => cred.input_secret.amount,
+            InputMaterials::Presigned(input) => Amount {
+                value: input.pseudo_output_amount.value,
+                token_id: input.pseudo_output_amount.token_id.into(),
```

Comment:
> Does it make sense to make a contingent input optionally blinded if we find a useful case for it?

I think, we should figure out the use-case for this before writing code that allows it and does it, because extra possibilities makes the code more complex.

Right now the only thing that a signed contingent input can be used for is to add it to the transaction builder. There is no way that an MLSAG can be added to a transaction which will be valid if we don't know the value, token id, and blinding factor of the pseudo-output that it produces.

Moreover, if the counter party has no insight into the value and token id, then there's no way that they can build a balanced transaction. So it's hard to see how any system could work that conceals this.

If the value and token id are known, then there doesn't seem to be anything lost by also revealing the blinding factor, since concealing those is its purpose.